File: Deprecated/unix/ehr_geometry/curvature3D.cpp
#include "curvature3D.h"

#include <stdio.h>

typedef map<TriPosition, Curvature3D*, TriPositionCompare> Curvature3DIndex;
static Curvature3DIndex* Index = NULL;

Curvature3D::Curvature3D( Vertex& v ){    
  sectionalCurvs = new vector<GeoQuant*>();
  partials = new vector<GeoQuant*>();

  GeoQuant* sectC;
  GeoQuant* partial;
    
  for(int i = 0; i < v.getLocalEdges()->size(); i++){
    sectC = SectionalCurvature::At(Triangulation::edgeTable[(*(v.getLocalEdges()))[i]]);
    partial = PartialEdge::At(v, Triangulation::edgeTable[(*(v.getLocalEdges()))[i]]);
    sectC->addDependent(this);
    partial->addDependent(this);
    sectionalCurvs->push_back( sectC );
    partials->push_back( partial );
  }
}

void Curvature3D::recalculate() {
  double curv = 0; 
  GeoQuant* sectC;
  GeoQuant* partial;
  for(int ii = 0; ii < sectionalCurvs->size(); ii++){
    sectC = sectionalCurvs->at(ii);
    partial = partials->at(ii);
    curv += sectC->getValue()* partial->getValue();
  }
    
  value = curv;
}

Curvature3D::~Curvature3D() {}

void Curvature3D::remove() {
     deleteDependents();
     for(int ii = 0; ii < sectionalCurvs->size(); ii++){
       sectionalCurvs->at(ii)->removeDependent(this);
       partials->at(ii)->removeDependent(this);
     }    
     Index->erase(pos);
     delete sectionalCurvs;
     delete partials;
     delete this;
}

Curvature3D* Curvature3D::At( Vertex& v  ){
  TriPosition T( 1, v.getSerialNumber() );
  if( Index == NULL ) Index = new Curvature3DIndex();
  Curvature3DIndex::iterator iter = Index->find( T );

  if( iter == Index->end() ){
    Curvature3D* val = new Curvature3D( v );
    val->pos = T;
    Index->insert( make_pair( T, val ) );
    return val;
  } else {
    return iter->second;
  }
}


void Curvature3D::CleanUp(){
  if( Index == NULL ) return;
  Curvature3DIndex::iterator iter;
  Curvature3DIndex copy = *Index;
  for(iter = copy.begin(); iter != copy.end(); iter++) {
    iter->second->remove();
  }
    
  delete Index;
  Index = NULL;
}

void Curvature3D::Record( char* filename ){
  FILE* output = fopen( filename, "a+" );

  Curvature3DIndex::iterator iter;
  for(iter = Index->begin(); iter != Index->end(); iter++)
    fprintf( output, "%lf ", iter->second->getValue() );
  fprintf( output, "\n");

  fclose( output );
}

File: trunk/Java/src/Geoquant/Curvature3D.java
package Geoquant;

import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedList;

import Triangulation.Edge;
import Triangulation.Triangulation;
import Triangulation.Vertex;

public class Curvature3D extends Geoquant {
  // Index map
  private static HashMap<TriPosition, Curvature3D> Index = new HashMap<TriPosition, Curvature3D>();
  
  // Needed geoquants
  private LinkedList<SectionalCurvature> sec_curvs;
  private LinkedList<PartialEdge> partials;
  
  public Curvature3D(Vertex v) {
    super();
    sec_curvs = new LinkedList<SectionalCurvature>();
    partials = new LinkedList<PartialEdge>();
    SectionalCurvature sc;
    PartialEdge pe;
    for(Edge e : v.getLocalEdges()) {
      sc = SectionalCurvature.At(e);
      sc.addDependent(this);
      pe = PartialEdge.At(v, e);
      pe.addDependent(this);
      sec_curvs.add(sc);
      partials.add(pe);
    }
  }
  
  protected void recalculate() {
    value = 0;
    Iterator<PartialEdge> pe_it = partials.iterator();
    for(SectionalCurvature sc : sec_curvs) {
      value += sc.getValue() * pe_it.next().getValue();
    }
  }
 
  protected void remove() {
    deleteDependents();
    for(SectionalCurvature sc : sec_curvs) {
      sc.removeDependent(this);
    }
    for(PartialEdge pe : partials) {
      pe.removeDependent(this);
    }
    sec_curvs.clear();
    partials.clear();
    Index.remove(pos);
  }
  
  public static Curvature3D At(Vertex v) {
    TriPosition T = new TriPosition(v.getSerialNumber());
    Curvature3D q = Index.get(T);
    if(q == null) {
      q = new Curvature3D(v);
      q.pos = T;
      Index.put(T, q);
    }
    return q;
  }
  
  public static double valueAt(Vertex v) {
    return At(v).getValue();
  }
    
  public static class Sum extends Geoquant {
    LinkedList<Curvature3D> curvs = new LinkedList<Curvature3D>();
    
    Sum() {
      super();
      Curvature3D k;
      for(Vertex v : Triangulation.vertexTable.values()) {
        k = Curvature3D.At(v);
        k.addDependent(this);
        curvs.add(k);
      }
    }
    protected void recalculate() {
      value = 0;
      for(Curvature3D k : curvs) {
        value += k.getValue();
      }
    }

    protected void remove() {
      deleteDependents();
      for(Curvature3D k : curvs) {
        k.removeDependent(this);
      }
      curvs.clear();
    }
  }

}
